apis/Google.Cloud.Logging.NLog/Google.Cloud.Logging.NLog/GoogleStackdriverTarget.cs

```diff
@@ -180,13 +95,63 @@ protected override void InitializeTarget()
             GaxPreconditions.CheckNotNullOrEmpty(logId, nameof(LogId));
 
+            if (SendJsonPayload)
+            {
+                if (JsonConvertFunction != null)
+                {
+                    // Use the function provided directly.
+                    GaxPreconditions.CheckState(
+                        string.IsNullOrWhiteSpace(JsonConvertTypeName) && string.IsNullOrWhiteSpace(JsonConvertMethodName),
+                        $"{nameof(JsonConvertTypeName)} and {nameof(JsonConvertMethodName)} must not be set along with {nameof(JsonConvertFunction)}.");
+                    _jsonConvertFunction = JsonConvertFunction;
+                }
+                else if (!string.IsNullOrWhiteSpace(JsonConvertTypeName) || !string.IsNullOrWhiteSpace(JsonConvertMethodName))
+                {
+                    // Use the method refered to by type-name and method-name.
+                    GaxPreconditions.CheckState(
+                        !string.IsNullOrWhiteSpace(JsonConvertTypeName) && !string.IsNullOrWhiteSpace(JsonConvertMethodName),
+                        $"Either both or neither of {nameof(JsonConvertTypeName)} and {nameof(JsonConvertMethodName)} must be specified.");
+                    _jsonConvertFunction = BuildAndVerifyJsonConverter();
+                }
+                else
+                {
+                    // Use default json.net based converter.
+                    _jsonConvertFunction = o => ProtoConverter.Convert(o is null ? null : JToken.FromObject(o));
+                }
+            }
+
             ActivateLogIdAndResource(logId);
 
             _client = _client ?? BuildLoggingServiceClient();
 
             base.InitializeTarget();
         }
 
+        private Func<object, Value> BuildAndVerifyJsonConverter()
+        {
+            var type = System.Type.GetType(JsonConvertTypeName, throwOnError: false);
+            GaxPreconditions.CheckState(type != null, "A type with the specified name cannot be found: '{0}'", JsonConvertTypeName);
+            var methodInfo = type.GetTypeInfo()
+                .GetMethods(BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance)
+                .Where(m => m.Name == JsonConvertMethodName && !m.IsAbstract && !m.IsGenericMethod &&
+                    m.GetParameters().Length == 1 && m.GetParameters()[0].ParameterType == typeof(object) &&
+                    (m.ReturnType == typeof(Value) || typeof(JToken).IsAssignableFrom(m.ReturnType)))
+                .FirstOrDefault();
+            GaxPreconditions.CheckState(methodInfo != null,
+                $"A suitable public method named '{JsonConvertMethodName}' cannot be found in type '{JsonConvertTypeName}'. " +
+                "The public method must have a single parameter of type 'object', and a return type of 'Value' or 'JToken'");
+            object instance = methodInfo.IsStatic ? null : Activator.CreateInstance(type);

```

Comment:
If this throws, are we happy that the exception is just thrown directly, or do you think we should wrap it with a message explaining what's going on?

Reply:
Done